File: src/prompt-poker.go
package main

import (
	"fmt"
	"os"
)

var usage = fmt.Sprintf("USAGE: %s FILEPATH\n", os.Args[0])

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, usage)
		os.Exit(2)
	}
	if err := createWriteChmodFile(os.Args[1]); err != nil {
		fmt.Fprintf(os.Stderr, "%v\n", err)
		os.Exit(1)
	}
}

func createWriteChmodFile(filepath string) error {
	// create
	f, err := os.OpenFile(filepath, os.O_WRONLY|os.O_CREATE, 0o600)
	if err != nil {
		return err
	}
	defer f.Close()
	// write
	if _, err := f.WriteString("hello prompting\n"); err != nil {
		return err
	}
	// chmod
	if err := f.Chmod(0o644); err != nil {
		return err
	}
	if err := f.Sync(); err != nil {
		return err
	}
	// write again
	if _, err := f.WriteString("hello again\n"); err != nil {
		return err
	}
	return nil
}

File: src/prompt-poker.rs
use std::env;
use std::fs;
use std::io;
use std::io::Write;
use std::os::unix::fs::{OpenOptionsExt, PermissionsExt};
use std::process;

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("USAGE: {} FILEPATH", args[0]);
        process::exit(2);
    }
    if let Err(err) = create_write_chmod_file(&args[1]) {
        eprintln!("{}", err);
        process::exit(1);
    }
}

fn create_write_chmod_file(filepath: &str) -> io::Result<()> {
    // create
    let mut file = fs::File::options()
        .create(true)
        .write(true)
        .mode(0o600)
        .open(filepath)?;
    // write
    file.write_all(b"hello prompting\n")?;
    // chmod
    let mut perms = file.metadata()?.permissions();
    perms.set_mode(0o644);
    file.set_permissions(perms)?;
    // sync
    file.sync_all()?;
    // write again
    file.write_all(b"hello again\n")?;
    Ok(())
}
